proposed/2023/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,199 @@
+# ***Release staging and deprecation***
+<!-- Replace `Title` with an appropriate title for your design -->
+
+- Author Name Artak Mkrtchyan ([mkArtakMSFT](https://github.com/mkArtakMSFT))
+- GitHub Issue <https://github.com/NuGet/NuGetGallery/issues/3931>
+
+## Summary
+
+This proposal introduces a concept for grouping packages that are published to NuGet.
+This will allow one to easily identify all the packages released as part of a single group as well as request bulk-processing
+ for all packages which belong to the same group. For example, all packages published as part of .NET 8 release will be staged
+ together and then also can be deprecated together by a single call to NuGet APIs.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set.
+ Currently, NuGet provides no functionality for definding / grouping packages into such sets.
+ This results in package owners having to keep track of these package sets on their own.
+ There are three specific sets of scenarios which this feature will simplify dramatically:
+1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages,
+ so that the release will consist only of a single action of marking the set of packages as `published`,
+ which would result all the packages in the set to be marked as published at the same time.
+2. Sometimes developers have to take a dependency on several related packages in their projects, which are part of the same "release".
+ Without having a concept of the `release` developers have to guess which versions of those packages are compatible.
+ As a result, most of the time developers would simply pick the latest version of all these packages, which sometimes can lead to incompatibility issues down the road.
+3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy.
+Hence, package owners try to deprecate packages which are out of support. Without ability to group packages,
+ package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+A new NuGet CLI command `stage` will be added, which will allow a user to stage a package to nuget, specifying a `group` parameter.
```

Comment:
Could `stage` be just an option for existing `push` command? like `dotnet nuget push  my.nupkg --stage --group mygroup`

Reply:
I'm afraid of the risk here. Imagine a person tries to stage a package, but accidentally mistypes the command and ends up actually making it available to the public.
This risk is not small because of two reasons:
1. `staging` should be treated as a disposable concept. So most of the packages that will be staged, will eventually be removed - freeing up space from NuGet servers, leaving no trace of these packages on the server. Because staging for a release will happen in groups, imagine accidentally not staging, but publishing a whole set of packages to NuGet. Removing packages, as I understand, is not possible today (unlisting is, but it's not the same).
2. What if a package which was meant to be staged only, contains a security patch?

Reply:
Currently symbol push and nupkg push are both lumped under `dotnet nuget push` so it seems like that's the place to expand the "push" concept rather than adding a new verb (i.e. we didn't add a `dotnet nuget symbol-push` at the time, we overloaded the existing concept. 

That being said, it's under the `nuget` so having `dotnet nuget stage` doesn't seem so bad either. 

I think having it as part of `push` is a bit better since all of the options on `push` right now probably make sense for `stage` (see `dotnet nuget push --help`). So if we had a separate verb we'd be copying all of those options over or having to make the decision of whether or not to copy it over. Easier to overload IMO.

Reply:
Another point is that the `dotnet nuget delete` gesture (the buddy of `dotnet nuget push`) would need another new verb like `unstage` to pair with `stage`. At that point it feels like too many new verbs.

Reply:
It might be worth alluding to the fact that not all sources will implement staging immediately or every. The `--stage` option (or `stage` verb) should fail gracefully when stage isn't supported. At the implementation level this means adding a way for sources to advertise staging support, which is a solve problem with our V3 service index model -- servers can advertise capabilities and client can enrich the error/fallback experience in that case.

Reply:
>Another point is that the dotnet nuget delete gesture (the buddy of dotnet nuget push) would need another new verb like unstage to pair with stage. At that point it feels like too many new verbs.

Alternatively, the --stage option can be applicable to the `dotnet nuget delete` option, so there is no extra word involved.